### PR TITLE
Improve the Gov Module Mechanism by BLV Labs

### PR DESCRIPTION
**Pull request by TuSoict**
Spec or proposal link: https://commonwealth.im/terra-luna-classic-lunc/discussion/24630-proposal-to-improve-the-gov-module-mechanism-by-blv-labs 

---
sidebar_position: 1
---

# `x/gov`
# Custom Governace Module Updates

## Concepts

*Disclaimer: This is work in progress. Mechanisms are susceptible to change.*

The governance process is divided in a few steps that are outlined below:
* **Proposal submission:** Proposal is submitted to the blockchain with a
  deposit.
* **Vote:** When the deposit for a proposal reaches a minimum threshold equivalent to
 500 USD (`MinUusdDeposit`), proposal is confirmed and vote opens. 
* **Execution** After a period of time, the votes are tallied and depending
  on the result, the messages in the proposal will be executed.
#### Right to submit a proposal

Every account can submit proposals by sending a `MsgSubmitProposal` transaction.
Once a proposal is submitted, it is identified by its unique `proposalID`.
#### Proposal Messages

A proposal includes an array of `sdk.Msg`s which are executed automatically if the
proposal passes. The messages are executed by the governance `ModuleAccount` itself. Modules
such as `x/upgrade`, that want to allow certain messages to be executed by governance
only should add a whitelist within the respective msg server, granting the governance
module the right to execute the message once a quorum has been reached. The governance
module uses the `MsgServiceRouter` to check that these messages are correctly constructed
and have a respective path to execute on but do not perform a full validity check.
### Deposit

To prevent spam, proposals must be submitted with a deposit in the coins defined by the equation `luncMinDepositAmountBasedOnUusd` = `MinUusdDeposit` / real-time price of LUNC at the time of proposal submission.
```keeper reference
github.com/classic-terra/core/v3/custom/gov/keeper/proposal.go#L136-L149
```
The value of `luncMinDepositAmountBasedOnUusd` will be stored in KVStores, using a key defined as `UUSDMinKeyPrefix|proposalID`, where the proposal ID is appended to the prefix, represented as a single byte.
```types reference
github.com/classic-terra/core/v3/custom/gov/types/keys.go#L20-L22
```

When a proposal is submitted, it has to be accompanied with a deposit that must be
strictly positive, but can be inferior to `luncMinDepositAmountBasedOnUusd`. The submitter doesn't need
to pay for the entire deposit on their own. The newly created proposal is stored in
an *inactive proposal queue* and stays there until its deposit passes the `luncMinDepositAmountBasedOnUusd`.
Other token holders can increase the proposal's deposit by sending a `Deposit`
transaction. If a proposal doesn't pass the `luncMinDepositAmountBasedOnUusd` before the deposit end time
(the time when deposits are no longer accepted), the proposal will be destroyed: the
proposal will be removed from state and the deposit will be burned (see x/gov `EndBlocker`).
When a proposal deposit passes the `luncMinDepositAmountBasedOnUusd` threshold (even during the proposal
submission) before the deposit end time, the proposal will be moved into the
*active proposal queue* and the voting period will begin.

The deposit is kept in escrow and held by the governance `ModuleAccount` until the
proposal is finalized (passed or rejected).
#### Deposit refund and burn

When a proposal is finalized, the coins from the deposit are either refunded or burned
according to the final tally of the proposal:

* If the proposal is approved or rejected but *not* vetoed, each deposit will be
  automatically refunded to its respective depositor (transferred from the governance
  `ModuleAccount`).
* When the proposal is vetoed with greater than 1/3, deposits will be burned from the
  governance `ModuleAccount` and the proposal information along with its deposit
  information will be removed from state.
* All refunded or burned deposits are removed from the state. Events are issued when
  burning or refunding a deposit.
#### Voting period

Once a proposal reaches `luncMinDepositAmountBasedOnUusd`, it immediately enters `Voting period`. We
define `Voting period` as the interval between the moment the vote opens and
the moment the vote closes. `Voting period` should always be shorter than
`Unbonding period` to prevent double voting. The initial value of
`Voting period` is 2 weeks.

## Stores

:::note
Stores are KVStores in the multi-store. The key to find the store is the first parameter in the list
:::
We will use one KVStore `Governance` to store five mappings:

* A mapping from `proposalID|'proposal'` to `Proposal`.
* A mapping from `proposalID|'addresses'|address` to `Vote`. This mapping allows
  us to query all addresses that voted on the proposal along with their vote by
  doing a range query on `proposalID:addresses`.
* A mapping from `ParamsKey|'Params'` to `Params`. This map allows to query all 
  x/gov params.
* A mapping from `VotingPeriodProposalKeyPrefix|proposalID` to a single byte. This allows
  us to know if a proposal is in the voting period or not with very low gas cost.
* A mapping from `UUSDMinKeyPrefix|proposalID` to a single byte. This allows us to determine the minimum amount of LUNC that a specific proposal must deposit to pass the deposit phase.
  
For pseudocode purposes, here are the two function we will use to read or write in stores:

* `load(StoreKey, Key)`: Retrieve item stored at key `Key` in store found at key `StoreKey` in the multistore
* `store(StoreKey, Key, value)`: Write value `Value` at key `Key` in store found at key `StoreKey` in the multistore

## Parameters

The governance module contains the following parameters:

| Key                           | Type             | Example                                 |
|-------------------------------|------------------|-----------------------------------------|
| min_deposit       | array (coins)    | [{"denom":"uatom","amount":"10000000"}] |
| max_deposit_period            | string (time ns) | "172800000000000" (17280s)              |
| voting_period                 | string (time ns) | "172800000000000" (17280s)              |
| quorum                        | string (dec)     | "0.334000000000000000"                  |
| threshold                     | string (dec)     | "0.500000000000000000"                  |
| veto                          | string (dec)     | "0.334000000000000000"                  |
| burn_proposal_deposit_prevote | bool             | false                                   |
| burn_vote_quorum              | bool             | false                                   |
| burn_vote_veto                | bool             | true                                    |
| min_uusd_deposit              | coins            |{"denom":"uusd","amount":"500000"}       |
Currently, the default value of `min_uusd_deposit` is set to `500` usd, but this value can be 
updated by the community in the future.

### CLI

A user can query and interact with the `gov` module using the CLI.
We have added two new commands 

##### MinimalDeposit
The `MinimalDeposit` command enables users to query the minimum LUNC deposit required for a specific proposal, calculated based on the UUSD value. 

```bash
terrad q gov min-deposit [proposal-id] 
```
Example:

```bash
terrad q gov min-deposit 1 cosmos1..
```

##### CustomParams
The `CustomParams` command allows users to query the custom parameters of the module.

```bash
terrad q gov custom-params 
```
Example:

```bash
terrad q gov custom-params
```
## Future Improvements
We have successfully upgraded the chain by increasing the consensus version to 5. Moving forward,
 should the community require updates related to these matters, the BLV Team will be available to 
 assist with the implementation process.

### CODE REVIEW

**Comment by duonghb53** [comment]
> Did you account for the minimum initial deposit?
> 
> On classic we have a customized minimum deposit that needs to be put into the proposal directly on creation. This might lead to an issue where the UI/client doesn't know what the necessary initial deposit is as the min deposit is only stored after creating the proposal.
> 
> Maybe it would make more sense to not make it a "real-time" adjustment but a recurring event (e.g. each X blocks) that sets the actual minimum deposit parameter internally? This would offer the most backward compatible approach maybe? Just a thought.

@StrathCole 
I can add a query to calculate the minimum `lunc` according to `uusd` at the time a user wants to create a proposal. Then, the UI or client can query at each block interval once. If the proposal is submitted within that same block, the price of uusd will not change, so the minimum deposit calculated will be the same.

**Comment by TuSoict** [comment]
> Some tests seem to be failing. Please check those before requesting a review on this PR.
> 
> Additional note: In my opinion, you will need a further parameter or hard-coded value of a grace-period. If the value of LUNC changes real-time then the querying of the min deposit is already outdated when submitting the actual deposit. This would probably lead to confusion/frustration. In case you already accounted for that or are storing the actual necessary LUNC value in a store, then it's fine. I didn't check the code yet.

Hello StrathCole
Thanks for your comment

All tests passed.
Please review our PR code

The amount of LUNC to be deposited will be calculated during the time the user makes the proposal and saved to the store by the key proposalID
Each proposal will require a different amount of deposit based on the price of Lunc at the time the proposal is created

https://github.com/BLV-Labs/core/blob/d780c04650bef3fbf68704821fd079cbc7235a0d/custom/gov/keeper/proposal.go#L97

**Comment by TuSoict** [comment]
> > If the proposal is submitted within that same block, the price of uusd will not change, so the minimum deposit calculated will be the same.
> 
> @duonghb53 That is unlikely to happen unless you use cli, I guess. Filling in the fields in the UI to make the proposal and processing it after clicking "confirm" will certainly not happen in the same block. At least it will take 2 blocks maybe more depending on how fast you put in your wallet password etc.

Hi StrathCole,
I understand your concern.

But I want to know that the LUNC deposit amount will only be counted when the user makes confirm - that is, submits the proposal to the on-chain, but nothing is done when creating or drafting the proposal.

That means in terms of UI/UX, it will be counted when the user presses the "Confirm" button.

https://github.com/BLV-Labs/core/blob/d780c04650bef3fbf68704821fd079cbc7235a0d/custom/gov/keeper/keeper.go#L95

And each proposal will have a different deposit amount.

We have also provided an additional CLI that helps users determine the amount of LUNC they need to deposit according to their proposal ID.

This will not be confusing for users when using.

**Comment by TuSoict** [comment]
> > That means in terms of UI/UX, it will be counted when the user presses the "Confirm" button.
> 
> This is what I said. Maybe I didn't explain well. When you click confirm, then the wallet, e.g. Keplr, pops up. Then you have to review and confirm the transaction in the wallet. Depending on the wallet and settings this is done by bio auth or password. This process will certainly take longer than the block end of where the confirm was pressed. So what will be actually sent to the chain will be outdated by 2 blocks at least (approx 10-12 seconds) most of the time. If the user takes a bit longer to confirm then even more.

Hi @StrathCole  
I guess you haven't read the code from our PR.

But I will try to explain it to you

1. The community completely agrees that the Proposal should keep the value at $500 (in USD) regardless of LUNC price going up or down

LUNC price is always public on many websites/exchanges (Binance/coinmarketcap, coingecko), 
if you have tried to check the oracle module, the LUNC price from Oracle is also close to this price
from there, any LUNC user who wants to make a proposal for LUNC on the chain can easily calculate how much LUNC to send for their proposal before they send.

_requiredAmount = $500/(current LUNC price UUSD base)_

of course, it may not be the exact number, it may be a little bit off after they send. But this is normal, at least they know the approximate number before they start.

so I think it is really clear to users, no confusion.

**2. Proposal Creation Process Analysis**
Users will need to create 2 steps:

S1. Submit a proposal with their information. In this step, users will not need to deposit the entire LUNC amount, 

The proposal after creation will be pushed to the deposit stage

S2. Deposit: Users need to deposit to activate their proposal. And this is the step where they need to deposit enough money to exceed the limit to push the proposal to the voting stage.

And here, it is important that we have chosen the time when users submit a proposal to calculate the amount of LUNC needed to deposit based on the current LUNC price and keep it at $500.

And each proposal when submitted, the amount of LUNC needed to deposit will be stored, ensuring fairness for users.

Soon after, we provide a new CLI, which helps users see the exact amount of LUNC to deposit by proposalID

and once the exact amount is available, users will continue to deposit LUNC and activate their proposal.

**_In short, compared to the previous conventional way, LUNC Users will only need to change:
When creating a proposal, instead of depositing a fixed amount of 5M (LUNC), they now need to deposit an amount of LUNC with a convertible value of $500._**

**Comment by TuSoict** [comment]
> > > > That means in terms of UI/UX, it will be counted when the user presses the "Confirm" button.
> > > 
> > > 
> > > This is what I said. Maybe I didn't explain well. When you click confirm, then the wallet, e.g. Keplr, pops up. Then you have to review and confirm the transaction in the wallet. Depending on the wallet and settings this is done by bio auth or password. This process will certainly take longer than the block end of where the confirm was pressed. So what will be actually sent to the chain will be outdated by 2 blocks at least (approx 10-12 seconds) most of the time. If the user takes a bit longer to confirm then even more.
> > 
> > 
> > Hi @StrathCole I guess you haven't read the code from our PR.
> > But I will try to explain it to you
> > […]
> > **_In short, compared to the previous conventional way, LUNC Users will only need to change: When creating a proposal, instead of depositing a fixed amount of 5M (LUNC), they now need to deposit an amount of LUNC with a convertible value of $500._**
> 
> No, this is not what I am referring to. I know what you did by scrolling through the code at least. What I am referring to is the **initial** deposit necessary to create a proposal at all. This is a percentage of the full deposit amount, currently 20%. So users need to deposit 1M currently or 100$ with your method. But the creation transaction will fail if you provide below that amount, so if the price of LUNC lowers a tiny bit from submitting until confirming the tx … the tx will fail.

You mean "minInitialDepositRatio‎" right?

if yes, i think you haven't read our code
we also have this logic and update according to new parameter

read here:
https://github.com/BLV-Labs/core/blob/4c44c75cdd7ab36c281e9ad178d6e1cb7466332c/custom/gov/keeper/deposit.go#L103

https://github.com/BLV-Labs/core/blob/4c44c75cdd7ab36c281e9ad178d6e1cb7466332c/custom/auth/ante/min_initial_deposit.go#L56

It ensures the logic as what you said, but based on new parameter, User doesn't need to deposit full amount but they have to deposit initial amount with **minInitialDepositRatio‎**

Please read our code first.

**Comment by TuSoict** [comment]
> > No, this is not what I am referring to. I know what you did by scrolling through the code at least. What I am referring to is the **initial** deposit necessary to create a proposal at all. This is a percentage of the full deposit amount, currently 20%. So users need to deposit 1M currently or 100$ with your method. But the creation transaction will fail if you provide below that amount, so if the price of LUNC lowers a tiny bit from submitting until confirming the tx … the tx will fail.

in addtion, as per current logic, even 20% or 100% if LUNC price dumps to near 0 then the value of the proposal is very very small around 0.1-0.2$, it can no longer hold 100$. So what do you think?
So do you think it can ensure spam proposal or not?

==> We have read this logic and continue to improve it by:
When creating a proposal, users need to deposit a small amount of 20% first but based on the value of 500$ regardless of LUNC price going up or down.

That means it always ensures the value of the proposal when created is 100$
This is more optimal.

I hope you have a clear view

**Comment by TuSoict** [comment]
> > > No, this is not what I am referring to. I know what you did by scrolling through the code at least. What I am referring to is the **initial** deposit necessary to create a proposal at all. This is a percentage of the full deposit amount, currently 20%. So users need to deposit 1M currently or 100$ with your method. But the creation transaction will fail if you provide below that amount, so if the price of LUNC lowers a tiny bit from submitting until confirming the tx … the tx will fail.
> 
> in addtion, as per current logic, even 20% or 100% if LUNC price dumps to near 0 then the value of the proposal is very very small around 0.1-0.2$, it can no longer hold 100$. So what do you think? So do you think it can ensure spam proposal or not?
> 
> ==> We have read this logic and continue to improve it by: When creating a proposal, users need to deposit a small amount of 20% first but based on the value of 500$ regardless of LUNC price going up or down.
> 
> That means it always ensures the value of the proposal when created is 100$ This is more optimal.
> 
> I hope you have a clear view

in this case, if needed we suggest that we provide a CLI that helps users query the required deposit amount when initial deposit
Of course it will be an inaccurate number when the LUNC price fluctuates

**Comment by fragwuerdig** [comment]
closed - this breaks the unfork attempts approved by governance
